File: freight-triage-polyglot-ledger-recovery/environment/reconcile/internal/statx/variance_milli.go
package statx

// VarianceMilli is the variance_milli kernel.
func VarianceMilli(series []int64) int64 {
	if len(series) == 0 {
		return 0
	}
	count := int64(len(series))
	total := int64(0)
	for _, value := range series {
		total += value
	}
	mean := FloorDiv(total*1000, count)
	accumulator := int64(0)
	for _, value := range series {
		delta := value*1000 - mean
		accumulator += delta * delta / 1000
	}
	return FloorDiv(accumulator, count)
}

File: freight-triage-polyglot-ledger-recovery/environment/intake/src/com/freight/stats/VarianceMilli.java
package com.freight.stats;

/** variance_milli kernel. */
public final class VarianceMilli implements StatKernel {

    @Override
    public String name() {
        return "variance_milli";
    }

    @Override
    public long apply(long[] series) {
        if (series.length == 0) {
            return 0L;
        }
        long count = series.length;
        long total = 0L;
        for (int i = 0; i < series.length; i++) {
            total += series[i];
        }
        long mean = StatSupport.floorDiv(total * 1000L, count);
        long accumulator = 0L;
        for (int i = 0; i < series.length; i++) {
            long delta = series[i] * 1000L - mean;
            accumulator += delta * delta / 1000L;
        }
        return StatSupport.floorDiv(accumulator, count);
    }
}
